engine/table/src/main/java/io/deephaven/engine/util/AbstractScriptSession.java

```diff
@@ -141,8 +151,9 @@ public synchronized final Changes evaluateScript(final String script, @Nullable
         final Changes diff;
         // retain any objects which are created in the executed code, we'll release them when the script session
         // closes
+        variableAccessLock.writeLock().lock();
```

Comment:
Let's take a few minutes today and consider the three locks in play?
 * UG exclusive/shared lock - exclusive lock is held for the duration of any executed script code
 * ScriptSession state read/write lock, preventing any (well, most) writes while anyone is actively reading
 * ScriptSession instance lock, appears only used to guard snapshots for scope changes. Probably can be folded into the new read/write lock.

External callers could already hold UG shared/exclusive locks when reading from QueryScope, so with the current code we could hit a lock inversion issue - take the write lock, wait to acquire the UG exclusive lock, then in another thread while holding a either UG lock try and acquire the read lock.

I think we can simply move the writelock inside `apply(() -> {...})`, plus as above remove the `synchronized` keyword usage?

Reply:
One more reply from me, then i'll defer to when we do a call: it isnt quite so simple to remove synchronized, since you can't upgrade a read lock to a write lock. If the whole write lock needs to be wider (encompass calls to takeSnapshot, observeScopeChanges, possibly createDiff), then the UG exclusive lock needs to also be correspondingly bigger as well.